### PR TITLE
[V1][Spec Decode] Relaxed acceptance for thinking-phase tokens (port of #22238)

### PR DESCRIPTION
**Pull request by chaeminlim-mb**
## Purpose

Port #22238's relaxed-thinking speculative acceptance into V1. Default behavior is unchanged. With `speculative_config.relaxed_thinking=True`, inside a reasoning span a greedy draft token is accepted when it is in the target top-K (K=3) and `draft_logit >= top1_logit + log(0.5)`; outside a reasoning span, acceptance stays strict argmax.

`relaxed_thinking` is the only new config field. The ratio (0.5) and top-K (3) are fixed module constants, not `--speculative-config` fields, and there is no new environment variable.

#22238 was auto-closed after 90 days of inactivity, not rejected on review; this revives that work on V1.

## How it works

The per-request "inside a thinking span" signal is sourced from the existing sample-layer `ThinkingBudgetStateHolder` — already built from `reasoning_config`, carried on `SamplingMetadata`, and read by the rejection sampler for the thinking-token-budget feature. When `relaxed_thinking` is enabled, the holder also tracks `in_think` for budget-less rows (a `-1` sentinel keeps those entries inert for the budget's end-token forcing). This is gated on `relaxed_thinking`, so reasoning deployments that don't use it keep their exact prior behavior.

At spec-decode verification, `rejection_sample` refreshes `in_think` from the committed output (batch-row order, last-boundary-wins), reads the boundary token ids from the holder, and passes a per-row `in_think` mask to `relaxed_thinking_sample_kernel`. Inside a thinking span the kernel accepts a draft token if it sits in the target top-K and its logit is within `log(0.5)` of the top-1 logit; otherwise it stays strict argmax. Accepting a reasoning boundary token truncates the rest of the speculative window so the next step re-verifies under the updated thinking state.

When `relaxed_thinking` is off, the sampler passes a sentinel (-2) for the boundary token ids and no mask, so the greedy path stays bit-for-bit unchanged. No scheduler / request / model-runner plumbing is added — the state lives entirely in the sample layer.

## Test Plan

```bash
python -m pytest tests/test_config.py -k thinking -q
python -m pytest tests/v1/sample/test_rejection_sampler.py -k "relaxed or holder" -q   # kernel cases need a GPU (Triton)
python -m pytest tests/v1/logits_processors/test_correctness.py -q                     # thinking-budget: no regression
```

## Test Result

CPU (any box, no GPU) — 8 passed:

- config (5): the `relaxed_thinking` guards — incompatible with synthetic rejection sampling, async scheduling, and CPU workers; requires single-token reasoning start/end markers; marks the V2 model runner unsupported.
- holder (2): with `relaxed_thinking` the budget holder tracks a budget-less row and `refresh_in_think` derives `in_think` from committed output; without `relaxed_thinking` the row is dropped (prior thinking-budget contract preserved).
- accept-rule oracle (1): a pure-torch reference mirror of the kernel rule, checked over the accept / below-floor-reject / strict-when-not-thinking / boundary-truncation cases (no GPU, no Triton, no env var).

GPU (Triton kernel) — 6 passed: the 5 kernel cases plus the holder-wiring case drive the real `relaxed_thinking_sample_kernel` on-device and each cross-checks the CPU oracle, so the kernel and oracle cannot silently drift. Verified on MI300 (ROCm).

No thinking-budget regression: `tests/v1/logits_processors/test_correctness.py` (40 tests, including the budget-less "drops row" case) passes.

ruff and py_compile: pass on the changed files.

### CODE REVIEW

**Comment by chaeminlim-mb** [comment]
Rebased onto current upstream/main and resolved the scheduler.py conflict in cdd0571624f006d475bcddc2c612edf67b793aab. The resolution keeps the relaxed-thinking scheduler state and preserves upstream's vllm_config.num_speculative_tokens initialization. Ran pre-commit on the PR files and py_compile on the 9 changed Python files.

**Comment by chaeminlim-mb** [comment]
Co-authored by @edwinlim0919 (Edwin Lim, MangoBoost) — credited in the commit trailers for this PR. Tagging him here so he is looped in and can follow the review.

File: tests/v1/attention/test_rocm_aiter_mla_mtp_split.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    assert AiterMLAMetadataBuilder._mtp_decode_query_len(config) == 4
+    assert not AiterMLAMetadataBuilder._split_uniform_mtp_decode(config)
+    assert AiterMLAMetadataBuilder._allow_uniform_mtp_decode(config)
+    assert (
+        AiterMLAMetadataBuilder.get_cudagraph_support(config, None)
+        == AttentionCGSupport.UNIFORM_BATCH
+    )
+
+
+@pytest.mark.parametrize("tp_size", [1, 8])
```
Comment (AndreasKaratzas) [inline tests/v1/attention/test_rocm_aiter_mla_mtp_split.py]: This will just error probably. It will try to find 8 GPUs and it will only be able to find one. Why do we need 8 GPUs for this? I mean we can map it on a 2 gpu agent and vector 1 and 2 tp sizes

File: tests/v1/attention/test_rocm_aiter_mla_mtp_split.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+def test_rocm_aiter_mtp_decode_uses_native_qlen_by_default(tp_size):
+    config = _mtp_config(tp_size=tp_size)
+
+    assert AiterMLAMetadataBuilder._mtp_decode_query_len(config) == 4
+    assert not AiterMLAMetadataBuilder._split_uniform_mtp_decode(config)
+    assert AiterMLAMetadataBuilder._allow_uniform_mtp_decode(config)
+    assert (
+        AiterMLAMetadataBuilder.get_cudagraph_support(config, None)
+        == AttentionCGSupport.UNIFORM_BATCH
+    )
```
Comment (AndreasKaratzas) [inline tests/v1/attention/test_rocm_aiter_mla_mtp_split.py]: What is the usefulness of this test?

File: tests/v1/attention/test_rocm_aiter_mla_mtp_split.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    assert AiterMLAMetadataBuilder._split_uniform_mtp_decode(config)
+    assert AiterMLAMetadataBuilder._allow_uniform_mtp_decode(config)
+    assert (
+        AiterMLAMetadataBuilder.get_cudagraph_support(config, None)
+        == AttentionCGSupport.UNIFORM_BATCH
+    )
+
+
+@pytest.mark.parametrize("tp_size", [1, 8])
```
Comment (AndreasKaratzas) [inline tests/v1/attention/test_rocm_aiter_mla_mtp_split.py]: same

File: tests/v1/attention/test_rocm_aiter_mla_mtp_split.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    config = _mtp_config(tp_size=tp_size, num_speculative_tokens=4)
+
+    assert AiterMLAMetadataBuilder._mtp_decode_query_len(config) == 5
+    assert AiterMLAMetadataBuilder._split_uniform_mtp_decode(config)
+    assert AiterMLAMetadataBuilder._allow_uniform_mtp_decode(config)
+    assert (
+        AiterMLAMetadataBuilder.get_cudagraph_support(config, None)
+        == AttentionCGSupport.UNIFORM_BATCH
+    )
+
```
Comment (AndreasKaratzas) [inline tests/v1/attention/test_rocm_aiter_mla_mtp_split.py]: same

File: tests/v1/attention/test_rocm_aiter_mla_mtp_split.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+def test_rocm_aiter_mla_without_mtp_keeps_single_only_decode():
+    config = SimpleNamespace(
+        speculative_config=None,
+        parallel_config=SimpleNamespace(tensor_parallel_size=1),
+    )
+
+    assert AiterMLAMetadataBuilder._mtp_decode_query_len(config) is None
+    assert not AiterMLAMetadataBuilder._split_uniform_mtp_decode(config)
+    assert not AiterMLAMetadataBuilder._allow_uniform_mtp_decode(config)
```
Comment (AndreasKaratzas) [inline tests/v1/attention/test_rocm_aiter_mla_mtp_split.py]: same

File: tests/v1/sample/test_rejection_sampler.py [hunk 6]
```diff
@@ -14,12 +17,21 @@
...
 )
 from vllm.v1.sample.sampler import Sampler, SamplerOutput
 from vllm.v1.spec_decode.metadata import SpecDecodeMetadata
 
 DEVICE_TYPE = current_platform.device_type
+RELAXED_REJECTION_SAMPLE_DEVICE = DEVICE_TYPE or "cpu"
+requires_relaxed_rejection_sample = pytest.mark.skipif(
+    not (
+        current_platform.is_cuda_alike()
+        or os.environ.get("TRITON_INTERPRET") == "1"
```
Comment (AndreasKaratzas) [inline tests/v1/sample/test_rejection_sampler.py]: ?? can you explain this env var?

**Review by AndreasKaratzas — COMMENTED** [review]
not sure that these tests have been carefully written

**Comment by edwinlim0919** [comment]
Hi, this PR is not included in MangoBoost's merge list for vllm, and can be safely closed.
